dll/win32/shell32/folders/CRecycleBin.cpp

```diff
@@ -750,26 +750,39 @@ HRESULT WINAPI CRecycleBin::GetDetailsOf(PCUITEMID_CHILD pidl, UINT iColumn, LPS
     TRACE("(%p, %p, %d, %p)\n", this, pidl, iColumn, pDetails);
     if (iColumn >= COLUMNS_COUNT)
         return E_FAIL;
-    pDetails->fmt = RecycleBinColumns[iColumn].fmt;
-    pDetails->cxChar = RecycleBinColumns[iColumn].cxChars;
+
     if (pidl == NULL)
+    {
+        pDetails->fmt = RecycleBinColumns[iColumn].fmt;
+        pDetails->cxChar = RecycleBinColumns[iColumn].cxChars;
         return SHSetStrRet(&pDetails->str, RecycleBinColumns[iColumn].column_name_id);
+    }
 
     if (iColumn == COLUMN_NAME)
         return GetDisplayNameOf(pidl, SHGDN_NORMAL, &pDetails->str);
 
     pFileDetails = _ILGetRecycleStruct(pidl);
+    if (!pFileDetails && FAILED_UNEXPECTEDLY(E_INVALIDARG))
+        return E_INVALIDARG;
     switch (iColumn)
     {
         case COLUMN_DATEDEL:
             FormatDateTime(buffer, MAX_PATH, &pFileDetails->DeletionTime);
             break;
         case COLUMN_DELFROM:
             pszBackslash = wcsrchr(pFileDetails->szName, L'\\');
-            Length = (pszBackslash - pFileDetails->szName);
-            memcpy((LPVOID)buffer, pFileDetails->szName, Length * sizeof(WCHAR));
+            if (!pszBackslash)
+            {
+                ERR("Filename '%ls' not a valid path?\n", pFileDetails->szName);
+                Length = 0;
+            }
+            else
+            {
+                Length = (pszBackslash - pFileDetails->szName);
+                memcpy((LPVOID)buffer, pFileDetails->szName, Length * sizeof(WCHAR));
+            }
             buffer[Length] = UNICODE_NULL;
-            if (buffer[0] && buffer[1] == L':' && !buffer[2])
+            if (!buffer[2] && buffer[0] && buffer[1] == L':')
```

Comment:
It would be better to start checking at index 0, because then you do not read data past the 'valid' string.
In this case it's a local buffer, but still you might be reading uninitialized data.

Reply:
The reason I did this is faster short circuit.

As to why this is safe and reasonable:

1) We are operating on `WCHAR buffer[MAX_PATH];`
2) In all sane cases, the first 3 characters of buffer are `X:\` or `X:`.
3) In most cases, the file was not in the root so `!buffer[2]` on `X:\` is instantly false. The new if should be faster.
4) In the root case, we have to check all 3 statements in the if but we had to do that anyway because we are looking for `X:`.

Reply:
Is this a performance bottleneck / a hot path?
As said before, I prefer easier to read / to understand code over micro-optimized (without profiling I presume) 'faster' code.

Reply:
Any code executed (once for each file) when enumerating and displaying items as part of the collaboration between DefView and IShellFolder is performance sensitive. For the Recycle Bin, especially so because it does not contain subfolders, all deleted items are presented in the same folder.

Reply:
You haven't actually tested the performance but are trying to sell your stuff here.

Reply:
> You haven't actually tested the performance

2000 empty text files in the bin (deleted in a subfolder of `c:\`).

My code, the fastest of 10 refresh runs was 1870343 QPC ticks. Old code, fastest was 1991225. This is in a VM so take everything with a grain of salt.

Reply:
Yes, sometimes the compiler or even the CPU can do some funny stuff.

Reply:
whindsaks acted on your whimpering already ... **still: Do not collapse this discussion-thread please!**